torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -449,27 +494,32 @@ void TensorPipeAgent::pipeWrite(
     const std::shared_ptr<tensorpipe::Pipe>& pipe,
     Message&& rpcMessage,
     std::vector<c10::DeviceIndex>&& devices,
+    std::shared_ptr<FullDeviceContext> ctx,
     std::function<void(const tensorpipe::Error&)> fn) noexcept {
   tensorpipe::Message tpMessage;
   TensorpipeWriteBuffers tpBuffers;
 
-  std::tie(tpMessage, tpBuffers) =
-      tensorpipeSerialize(std::move(rpcMessage), std::move(devices));
+
+  std::tie(tpMessage, tpBuffers) = tensorpipeSerialize(
+      std::move(rpcMessage), std::move(devices), ctx);
 
   pipe->write(
       std::move(tpMessage),
       [tpBuffers{
            std::make_shared<TensorpipeWriteBuffers>(std::move(tpBuffers))},
-       fn{std::move(fn)}](
+       fn{std::move(fn)},
+       ctx{std::move(ctx)}](
           const tensorpipe::Error& error, tensorpipe::Message /* unused */) {
+        ctx->recordTensors(tpBuffers->tensors);
```

Comment:
Nit: I believe this could be done in `tensorpipeSerialize` (which mirrors @pritamdamania87's comment above). If my understanding is correct, recording a tensor w/ the caching allocator just adds the stream to a list, and that list is only consumed when the tensor is destroyed. Hence recording at any moment before the tensor's destruction is equivalent.